x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
         }
     }
 
+    private final Map<IndicesAccessControlCacheKey, ListenableFuture<IndicesAccessControl>> indicesAccessControlLookup =
+        new ConcurrentHashMap<>();
+
+    private void buildIndicesAccessControlAsync(
+        String action,
+        AuthorizationInfo authorizationInfo,
+        ResolvedIndices resolvedIndices,
+        Metadata metadata,
+        ActionListener<IndexAuthorizationResult> listener) {
+
+        final ActionListener<IndicesAccessControl> wrappingListener =
+            listener.map((IndicesAccessControl indicesAccessControl) -> new IndexAuthorizationResult(true, indicesAccessControl));
```

Comment:
Nit, can we "cache" the `IndexAuthorizationResult` object instead of the indicesAccessControl?